#!/usr/bin/env Rscript
library(argparse)

# Parse command line arguments
parser <- ArgumentParser()

parser$add_argument("--patient", help="Patient identifier")
parser$add_argument("--samples", help="Patient sample names")
parser$add_argument("--mtxdirs", help="comma seperated list of matrix folders")
parser$add_argument("--outdir", help="output directory")
parser$add_argument("--genome_ver", help="Genome version (hg19, hg38)", default = "hg38")
parser$add_argument("--cores", help="number of cores to use", default=4)
parser$add_argument("--trans", help="Numbat HMM transmission probability", default=1e-5)
parser$add_argument("--gamma", help="Numbat overdispersion parameter in allele counts", default=20)
parser$add_argument("--min_cells", help="Numbat minimum number of cells for which an pseudobulk HMM will be run", default=20)
parser$add_argument("--min_LLR", help="Numbat minimum log-likelihood ratio threshold to filter CNVs by. ", default=50)

args <- parser$parse_args()

patient = args$patient
samples = args$samples
mtxdirs = args$mtxdirs
outdir = args$outdir
genome_ver = args$genome_ver
cores = args$cores
trans = args$trans
gamma = args$gamma
min_cells = args$min_cells
min_LLR = args$min_LLR

setwd(outdir)

# Load libraries
library(numbat)
library(Seurat)

# For multi-sample patients parse and load
samps <- strsplit(samples, ",")[[1]]
mtx_dirs <- strsplit(mtxdirs, ",")[[1]]

allele_files <- file.path(outdir, paste0(samps, "_allele_counts.tsv.gz"))

if (length(samps) > 1) { 
    count_mats <- lapply(mtx_dirs, Read10X)
    
    mmats <- c()
    for (i in 2:length(samps)) {
        mmats <- c(mmats, paste0('count_mats[[', i, ']]'))
    }
    # Parseable by merge function below
    mmats <- paste0("c(", paste(mmats, collapse = ","), ")")
    count_mat <- merge(x = count_mats[[1]], y = eval(parse(text=mmats)), add.cell.ids = samps, project = patient)
    
    df_allele <- c()
    for (i in 1:length(samps)) {
        df_allele_ <- read.table(file = allele_files[i], header = T, sep = "\t")
        df_allele_$sample <- samps[i]
        df_allele_$group <- patient
        df_allele <- rbind(df_allele, df_allele_)
    }
    
} else {
    # For single sample patients
    count_mat <- Read10X(mtx_dirs)
    df_allele <- read.table(file = allele_files, header = T, sep = "\t")
    df_allele$sample <- samps[1]
    df_allele$group <- patient
}

# Save input files
save(x = count_mat, file = "count_mat.rda")
save(x = df_allele, file = "df_allele.rda")

# Run numbat
out = run_numbat(
    count_mat = count_mat, # gene x cell integer UMI count matrix 
    lambdas_ref = ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
    df_allele = df_allele, # allele dataframe generated by pileup_and_phase script
    gtf = eval(parse(text = paste0("gtf_", genome_ver))), # provided upon loading the package
    genetic_map = eval(parse(text = paste0("genetic_map_", genome_ver))), # provided upon loading the package
    min_cells = min_cells,
    t = trans,
    max_iter = 2,
    min_LLR = min_LLR,
    init_k = 3,
    ncores = cores,
    plot = TRUE,
    out_dir = outdir
)

save(x = out, file = "numbat_out.rda")