#!/usr/bin/env Rscript
library(argparse)

# Parse command line arguments
parser <- ArgumentParser()

parser$add_argument("--patient", help="Patient identifier")
parser$add_argument("--samples", help="Patient sample names")
parser$add_argument("--mtxdirs", help="comma seperated list of matrix folders")
parser$add_argument("--outdir", help="output directory")
parser$add_argument("--genome_ver", help="Genome version (hg19, hg38) [default %(default)s]", default = "hg38")
parser$add_argument("--cores", help="number of cores to use [default %(default)s]", default=4, type = "integer")
parser$add_argument("--trans", help="Numbat HMM transmission probability [default %(default)s]", default=1e-5, type="double")
parser$add_argument("--gamma", help="Numbat overdispersion parameter in allele counts [default %(default)s]", default=20, type = "integer")
parser$add_argument("--min_cells", help="Numbat minimum number of cells for which an pseudobulk HMM will be run [default %(default)s]", default=20, type="integer")
parser$add_argument("--min_LLR", help="Numbat minimum log-likelihood ratio threshold to filter CNVs by. [default %(default)s]", default=50, type="integer")
parser$add_argument("--multi_allelic", help="Flag to run numbat in multi-allelic mode [default %(default)s]", action="store_true", default=TRUE)
parser$add_argument("--high_purity", help="Flag to detect and exclude regions of clonal deletions/LOH before running Numbat. Recommended for cell line data or high-purity tumors", action="store_true", default=FALSE)

args <- parser$parse_args()

patient = args$patient
samples = args$samples
mtxdirs = args$mtxdirs
outdir = args$outdir
genome_ver = args$genome_ver
ncores = args$cores
trans = args$trans
gamma = args$gamma
min_cells = args$min_cells
min_LLR = args$min_LLR
multi_allelic = args$multi_allelic
high_purity = args$high_purity

setwd(outdir)

# Load libraries
library(numbat)
library(Seurat)

# For multi-sample patients parse and load
samps <- strsplit(samples, ",")[[1]]
mtx_dirs <- strsplit(mtxdirs, ",")[[1]]

allele_files <- file.path(outdir, paste0(samps, "_allele_counts.tsv.gz"))

if (length(samps) > 1) {
    message("Multiple samples detected. Reading and merging data...")    
}

df_allele <- c()
count_mat <- c()
for (i in 1:length(samps)) {
    # Load count matrix and relabel cells by their sample
    count_mat_ <- Read10X(mtx_dirs[i])

    # Check for multiomic and only return gene expression
    if (length(count_mat_) > 1 & !is.null(names(count_mat_))) {
        print("Multiomic data detected. Keeping only gene expression...")
        count_mat_ <- count_mat_[['Gene Expression']]
    }

    # Paste sample names into cell_ids
    colnames(count_mat_) <- paste(samps[i], "_", colnames(count_mat_), sep = "")
    count_mat <- cbind(count_mat, count_mat_)

    # Read and add cols to the allele df
    df_allele_ <- read.table(file = allele_files[i], header = T, sep = "\t")
    df_allele_$cell <- paste(samps[i], "_", df_allele_$cell, sep = "")
    df_allele_$sample <- samps[i]
    df_allele_$group <- patient
    df_allele <- rbind(df_allele, df_allele_)
}


# Save input files
save(x = count_mat, file = "count_mat.rda")
save(x = df_allele, file = "df_allele.rda")

# Set multi_allelic flag
if (multi_allelic) {
    multi_allelic=TRUE
} else {
    multi_allelic=FALSE
}

# Detect clonal loh for high purity flagged samples
segs_loh = NULL
if (high_purity) {
    print("Running with high purity flag on!")
    bulk = get_bulk(
        count_mat = count_mat,
        lambdas_ref = ref_hca,
        df_allele = df_allele,
        gtf = gtf_hg38,
        genetic_map = genetic_map_hg38
    )

    segs_loh = bulk %>% detect_clonal_loh(t = 1e-4)
}

# Run numbat
out = run_numbat(
    count_mat = count_mat, # gene x cell integer UMI count matrix 
    lambdas_ref = ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
    df_allele = df_allele, # allele dataframe generated by pileup_and_phase script
    genome = genome_ver,
    min_cells = min_cells,
    t = trans,
    max_iter = 2,
    min_LLR = min_LLR,
    init_k = 3,
    ncores = ncores,
    plot = TRUE,
    out_dir = outdir,
    multi_allelic = multi_allelic,
    segs_loh = segs_loh
)

save(x = out, file = "numbat_out.rda")
